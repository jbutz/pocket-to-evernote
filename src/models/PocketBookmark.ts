export class PocketBookmark {
    public readonly href: string;
    public readonly title: string;
    public readonly timeAdded: Date;
    public readonly tags: string[];
    constructor(props: {
        href: string,
        title: string,
        timeAdded: string,
        tags?: string
    }) {
        this.href = props.href;
        this.title = props.title;
        this.timeAdded = new Date(Number.parseInt(props.timeAdded, 10) * 1000);
        this.tags = (props.tags || '').split(',').filter((value) => !!value);
    }
}