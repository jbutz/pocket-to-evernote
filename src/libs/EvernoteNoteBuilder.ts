import { PocketBookmark } from '../models/PocketBookmark';
import { create as createXml, fragment as createXmlFragment } from 'xmlbuilder2';
import { XMLBuilder } from 'xmlbuilder2/lib/interfaces';

export class EvernoteNoteBuilder {
    public build(pocketBookmark: PocketBookmark): string {
        const xmlObj = this.constructBookmarkNote(pocketBookmark);

        return xmlObj.end({prettyPrint: true});
    }

    private constructBookmarkNote(pocketBookmark: PocketBookmark): XMLBuilder {
        return createXml({ version: '1.0', encoding: 'UTF-8' })
            .ele('en-note')
            .dtd({ sysID: 'http://xml.evernote.com/pub/enml2.dtd' })
            .ele('div', { 'style': '--en-clipped-content:bookmark' })
            .ele('div', { style: 'font-size: 16px; display:block; min-width: 100%; ' })
            .ele('div')
            .import(this.constructNoteBody(pocketBookmark));
    }

    private constructNoteBody(pocketBookmark: PocketBookmark) {
        const bodyWrapper = createXmlFragment()
            .ele('div', { style: 'display:block;height:100%;' });
        bodyWrapper.import(this.constructBookmarkTitle(pocketBookmark));
        bodyWrapper.import(this.constructBookmarkBody(pocketBookmark));

        return bodyWrapper;
    }

    private constructBookmarkTitle(pocketBookmark: PocketBookmark) {
        return createXmlFragment()
            .ele('div', { 'style': 'font-size:14px;line-height:20px;font-weight:bold;padding-bottom:10px;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;', })
            .txt(pocketBookmark.title);
    }

    private constructBookmarkBody(pocketBookmark: PocketBookmark) {
        const bodyFlexContainer = createXmlFragment()
            .ele('div', { style: 'display:flex;flex-flow:row nowrap;place-content:stretch flex-start;align-items:flex-start;padding-top:15px;border-top:1px solid rgb(211, 214, 216);height:100%;' })
            .com('<en-media type="image/png" hash="ff26897593310722d6f07e6fcba357ef" width="146" height="125" style="max-width:180px;vertical-align:top;display:inline-block;"></en-media>');
        const flexContent = bodyFlexContainer.ele('div', { style: 'margin-left:15px;flex:1 1 auto;display:block;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;height:100%;' });
        flexContent.ele('div', { style: 'display:block;text-overflow:ellipsis;overflow:hidden;white-space:nowrap;margin-bottom:15px;' })
            .com('<en-media type="image/png" hash="346e09471362f2907510a31812129cd2" width="16" height="16" style="vertical-align:top;display:inline-block;max-width:16px;"></en-media>')
            .ele('a', { href: pocketBookmark.href, target: '_blank', style: 'margin-left:10px;' })
            .txt(pocketBookmark.href);
        flexContent.ele('div', { style: 'white-space:normal;font-size:12px;overflow:hidden;height:200px;line-height:150%;' })
            .txt(pocketBookmark.title);
        return bodyFlexContainer;
    }


}