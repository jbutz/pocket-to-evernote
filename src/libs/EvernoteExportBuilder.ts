import { DateTime } from 'luxon';
import { create as createXml, fragment as createXmlFragment } from 'xmlbuilder2';
import { XMLBuilder } from 'xmlbuilder2/lib/interfaces';
import { PocketBookmark } from '../models/PocketBookmark';
import { EvernoteNoteBuilder } from './EvernoteNoteBuilder';

const MAX_TITLE_LENGTH = 255;
export class EvernoteExportBuilder {
    constructor(private readonly evernoteNoteBuilder: EvernoteNoteBuilder) { }

    public buildExport(pocketBookmarks: PocketBookmark[]): string {
        const exportEl = createXml({ version: '1.0', encoding: 'UTF-8' })
            .ele('en-export', {
                'export-date': EvernoteExportBuilder.formatDate(new Date()),
                application: 'Pocket to Evernote',
                version: 'Pocket to Evernote'
            })
            .dtd({ sysID: 'http://xml.evernote.com/pub/evernote-export3.dtd' });

        pocketBookmarks.forEach((bookmark) => {
            exportEl.import(this.buildNoteElement(bookmark));
        });

        return exportEl.end({ prettyPrint: true });
    }

    private buildNoteElement(pocketBookmark: PocketBookmark): XMLBuilder {
        const noteEl = createXmlFragment()
            .ele('note');

        noteEl.ele('title')
            .txt(EvernoteExportBuilder.formatNoteTitle(pocketBookmark.title));

        noteEl.ele('content').dat(this.evernoteNoteBuilder.build(pocketBookmark));

        noteEl.ele('created').txt(EvernoteExportBuilder.formatDate(pocketBookmark.timeAdded));
        noteEl.ele('updated').txt(EvernoteExportBuilder.formatDate(pocketBookmark.timeAdded));
        const noteAttributesEl = noteEl.ele('note-attributes');
        noteAttributesEl.ele('source').txt('Pocket');
        noteAttributesEl.ele('source-url').txt(pocketBookmark.href);
        noteAttributesEl.ele('source-application').txt('Pocket to Evernote');

        pocketBookmark.tags.forEach((tag) => {
            noteEl.ele('tag').txt(tag);
        });

        return noteEl;
    }

    static formatDate(input: Date): string {
        return DateTime.fromJSDate(input).toUTC().toFormat('yyyyMMdd\'T\'HHmmss\'Z\'');
    }

    static formatNoteTitle(input: string): string {
        const trimmedTitle = input.trim();
        if(trimmedTitle.length > MAX_TITLE_LENGTH) {
            return `${trimmedTitle.substring(MAX_TITLE_LENGTH - 3)}...`;
        }
        return trimmedTitle;
    }
}