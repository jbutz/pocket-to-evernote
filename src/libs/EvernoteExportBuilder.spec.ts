import { PocketBookmark } from '../models/PocketBookmark';
import { EvernoteExportBuilder } from './EvernoteExportBuilder';
import { EvernoteNoteBuilder } from './EvernoteNoteBuilder';
jest.mock('./EvernoteNoteBuilder');

describe('EvernoteExportBuilder', () => {
    let noteBuilder: EvernoteNoteBuilder;
    beforeEach(() => {
        noteBuilder = new EvernoteNoteBuilder();
        (noteBuilder.build as unknown as jest.SpyInstance).mockReturnValue('<en-note />');
    });

    it('instantiates without error', () => {
        expect(() => {
            new EvernoteExportBuilder(noteBuilder);
        }).not.toThrow();
    });

    describe('formatTimestamp', () => {
        it.each([
            [1593093910000, '20200625T140510Z'],
            [1586522154050, '20200410T123554Z'],
            ['2020-04-30T23:59:59.999-0800', '20200501T075959Z'],
        ])('formats "%s"', (dateInitValue, expected) => {
            const inputDate = new Date(dateInitValue);
            expect(EvernoteExportBuilder.formatDate(inputDate)).toEqual(expected);
        });
    });

    describe('buildExport', () => {
        beforeAll(() => {
            jest.useFakeTimers('modern');
            jest.setSystemTime(new Date('2020-10-04T14:53:44Z'));
        });
        afterAll(() => {
            jest.useRealTimers();
        });

        it('builds an export for 1 bookmark', async () => {
            const exportBuilder = new EvernoteExportBuilder(noteBuilder);

            expect(exportBuilder.buildExport([
                new PocketBookmark({
                    href: 'https://example.com',
                    title: 'Example Bookmark',
                    timeAdded: '1593093910',
                    tags: ''
                })
            ])).toMatchSnapshot();
        });

        it('builds an export for 2 bookmark2', async () => {
            const exportBuilder = new EvernoteExportBuilder(noteBuilder);

            expect(exportBuilder.buildExport([
                new PocketBookmark({
                    href: 'https://example.com',
                    title: 'Example Bookmark',
                    timeAdded: '1593093910',
                    tags: ''
                }),
                new PocketBookmark({
                    href: 'https://example.com/new-url',
                    title: 'Example Bookmark Number 2',
                    timeAdded: '1503093910',
                    tags: 'example,two'
                })
            ])).toMatchSnapshot();
        });
    });
});