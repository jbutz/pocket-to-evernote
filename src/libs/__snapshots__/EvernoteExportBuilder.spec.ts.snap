// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvernoteExportBuilder buildExport builds an export for 1 bookmark 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE en-export SYSTEM \\"http://xml.evernote.com/pub/evernote-export3.dtd\\">
<en-export export-date=\\"20201004T145344Z\\" application=\\"Pocket to Evernote\\" version=\\"Pocket to Evernote\\">
  <note>
    <title>Example Bookmark</title>
    <content><![CDATA[<en-note />]]></content>
    <created>20200625T140510Z</created>
    <updated>20200625T140510Z</updated>
    <note-attributes>
      <source>Pocket</source>
      <source-url>https://example.com</source-url>
      <source-application>Pocket to Evernote</source-application>
    </note-attributes>
  </note>
</en-export>"
`;

exports[`EvernoteExportBuilder buildExport builds an export for 2 bookmark2 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE en-export SYSTEM \\"http://xml.evernote.com/pub/evernote-export3.dtd\\">
<en-export export-date=\\"20201004T145344Z\\" application=\\"Pocket to Evernote\\" version=\\"Pocket to Evernote\\">
  <note>
    <title>Example Bookmark</title>
    <content><![CDATA[<en-note />]]></content>
    <created>20200625T140510Z</created>
    <updated>20200625T140510Z</updated>
    <note-attributes>
      <source>Pocket</source>
      <source-url>https://example.com</source-url>
      <source-application>Pocket to Evernote</source-application>
    </note-attributes>
  </note>
  <note>
    <title>Example Bookmark Number 2</title>
    <content><![CDATA[<en-note />]]></content>
    <created>20170818T220510Z</created>
    <updated>20170818T220510Z</updated>
    <note-attributes>
      <source>Pocket</source>
      <source-url>https://example.com/new-url</source-url>
      <source-application>Pocket to Evernote</source-application>
    </note-attributes>
    <tag>example</tag>
    <tag>two</tag>
  </note>
</en-export>"
`;

exports[`EvernoteExportBuilder buildExport shortens bookmarktitles 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE en-export SYSTEM \\"http://xml.evernote.com/pub/evernote-export3.dtd\\">
<en-export export-date=\\"20201004T145344Z\\" application=\\"Pocket to Evernote\\" version=\\"Pocket to Evernote\\">
  <note>
    <title>123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012...</title>
    <content><![CDATA[<en-note />]]></content>
    <created>20200625T140510Z</created>
    <updated>20200625T140510Z</updated>
    <note-attributes>
      <source>Pocket</source>
      <source-url>https://example.com</source-url>
      <source-application>Pocket to Evernote</source-application>
    </note-attributes>
  </note>
</en-export>"
`;
